1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,189 @@
+
+## Specification
+
+A `dispose{}` block is required for these use cases.
+
+### Execution
+
+The `dispose{}` block will execute:
+
+1. After `end{}` during normal command operation, and
+2. During command disposal that occurs when a terminating exception is thrown during pipeline operations.
+    - If the dispose block has already been called (as it is after the `end{}` step), it will simply return immediately.
+
+### Behaviour
+
+The `dispose{}` block _does not_ guarantee output, but neither does it prevent it.
+
+- If output is received from the `dispose{}` block during normal command operation, it will be sent along the pipeline as per usual.
+- If output is received from the `dispose{}` block during a halting pipeline, it will be **discarded**.
+
+As a result, documentation should advise that `dispose{}` is _not intended_ for any standard output and the limitations of doing so.
+Additionally, all other PowerShell data streams are accessible from `dispose{}`, meaning there is no issue using any of the following streams from within `dispose{}`:
+
+- Error
+- Warning
+- Verbose
+- Debug
+- Information
+
+Terminating errors can also be thrown from a `dispose{}` block, which will skip the invocation of the rest of that `dispose{}` block, without preventing other `dispose{}` blocks from being executed.
+
+### Implementation Details
+
+- A new `dispose` keyword, indicating a `NamedBlockAst` similarly to `begin`, `process`, and `end`, which is valid in the same contexts as the existing named block keywords.
+  - This includes new token types and appropriate entries in the `Tokenizer`'s keyword tables.
+- Additional members and constructors for `ScriptBlockAst` to enable it to be recognised by the parser.
+- Additional methods in `DlrScriptCommandProcessor` and `CompiledScriptBlock` to enable the block to be executed at the correct time.
+  - Both types implement or inherit from a type that implements `IDisposable`, and as such the existing `Dispose()` methods can generally be utilised to run this block.
+- Changes to the behaviour of command execution in `CommandProcessorBase` to recognise when a function or script cmdlet is being executed, and appropriately call their `Dispose()` methods to invoke the dispose block.
+  - This can include changes to the existing command disposal behaviour, to enable commands to dispose their resources as soon as their tasks in the pipeline are completed, instead of waiting for the entire pipeline to complete.
+  - Also required with this is an additional layer of `try/finally` during `Dispose()` events as we are opening up the possibility for users to throw terminating errors _during_ a command's disposal, and as such we must ensure that disposal still completes appropriately.
+
```

Comment:
It appears that GitHub is now broken with Edge. Commenting works with Chrome.

From twitter @essentialexch (me): First, what happens with dispose{} when a script is interrupted (e.g., ctrl-c)? Second, can't a PS Class inherit from IDisposable? (In PS, not C#.) Or is that plumbing not implemented?

Reply from @vexx32 : 1/2 Ctrl+C behaviour is the same as terminating error behaviour; Dispose{} is executed, and all output is suppressed (though errors and other streams' data submitted from Dispose{} is passed along as normal).  2/2 PS Classes can inherit IDisposable, but a PowerShell function cannot. Creating a cmdlet from a PS class has fairly brutal caveats, chief among them perhaps is that it's exceedingly difficult / nigh impossible to properly export a cmdlet from a module if it's a PS Class.

Reply:
Updated the RFC to mention these points a bit more clearly. Thank you! 💖 😊